service.go

```diff
@@ -0,0 +1,541 @@
+}
+
+// newService creates a new threads network service.
+func newService(t *threads) (*service, error) {
+	ps, err := pubsub.NewGossipSub(
+		t.ctx,
+		t.host,
+		pubsub.WithMessageSigning(false),
+		pubsub.WithStrictSignatureVerification(false))
+	if err != nil {
+		return nil, err
+	}
+
+	s := &service{
+		threads: t,
+		pubsub:  ps,
+	}
+
+	for _, id := range t.Threads() {
+		go s.subscribe(id)
+	}
+
+	// @todo: ts.pubsub.RegisterTopicValidator()
+
+	return s, nil
+}
+
+// Push receives a record request.
+func (s *service) Push(ctx context.Context, req *pb.PushRequest) (*pb.PushReply, error) {
+	// Verify the request
+	reqpk, err := requestPubKey(req)
+	if err != nil {
+		return nil, err
+	}
+	err = verifyRequestSignature(req.Record, reqpk, req.Header.Signature)
+	if err != nil {
+		return nil, err
+	}
+
+	reply := &pb.PushReply{Ok: true}
+
+	// Unpack the record
+	var fkey []byte
+	if req.Header.FollowKey != nil {
+		fkey = req.Header.FollowKey
+	} else {
+		fkey = s.threads.FollowKey(req.ThreadID.ID, req.LogID.ID)
+	}
+	if fkey == nil {
+		return nil, fmt.Errorf("could not find follow key")
+	}
+
+	rec, err := recordFromProto(req.Record, fkey)
+	if err != nil {
+		return nil, err
+	}
+	x, err := s.threads.blocks.Blockstore().Has(rec.Cid())
+	if err != nil {
+		return nil, err
+	}
+	if x {
+		return reply, nil
+	}
+
+	// Check if this log already exists
+	logpk := s.threads.PubKey(req.ThreadID.ID, req.LogID.ID)
+	if logpk == nil {
+		var kid peer.ID
+		if req.Header.ReadKeyLogID != nil {
+			kid = req.Header.ReadKeyLogID.ID
+		}
+		lg, err := s.handleInvite(ctx, req.ThreadID.ID, req.LogID.ID, kid, rec)
+		if err != nil {
+			return nil, err
+		}
+
+		if lg != nil {
+			// Notify existing logs of our new log
+			invite, err := cbor.NewInvite([]thread.LogInfo{*lg}, true)
+			if err != nil {
+				return nil, err
+			}
+			_, _, err = s.threads.Add(
+				ctx,
+				invite,
+				tserv.AddOpt.Thread(req.ThreadID.ID),
+				tserv.AddOpt.KeyLog(req.LogID.ID))
+			if err != nil {
+				return nil, err
+			}
+		}
+	} else {
+		// Verify node
+		err = rec.Verify(logpk)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = s.threads.Put(ctx, rec, tserv.PutOpt.Thread(req.ThreadID.ID), tserv.PutOpt.Log(req.LogID.ID))
+	if err != nil {
+		return nil, err
+	}
+
+	return reply, nil
+}
+
+func (s *service) Pull(ctx context.Context, req *pb.PullRequest) (*pb.PullReply, error) {
+	recs, err := s.threads.pullLocal(ctx, req.ThreadID.ID, req.LogID.ID, req.Offset.Cid, int(req.Limit))
+	if err != nil {
+		return nil, err
+	}
+
+	pbrecs := &pb.PullReply{
+		Records: make([]*pb.Record, len(recs)),
+	}
+	for i, r := range recs {
+		pbrecs.Records[i], err = cbor.RecordToProto(ctx, s.threads.dagService, r)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return pbrecs, nil
+}
+
+// push a record to log addresses and thread topic.
+func (s *service) push(ctx context.Context, rec thread.Record, id thread.ID, lid peer.ID, settings *tserv.AddSettings) error {
```

Comment:
This is the client-side push method. Here we gather the addresses at which to deliver the request.